allennlp/nn/decoding/beam_search.py

```diff
@@ -35,10 +35,13 @@ def search(self,
                         if step_num == num_steps and keep_final_unfinished_states:
                             finished_states.append(next_state)
                         next_states.append(next_state)
+            # TODO(mattg): do this sort on the GPU, not on the CPU, and copy once.
             to_sort = [(-state.score.data[0], state) for state in next_states]
             to_sort.sort(key=lambda x: x[0])
             states = [state[1] for state in to_sort[:self._beam_size]]
-        finished_to_sort = [(-state.score.data[0], state) for state in finished_states]
+        # The time this one takes is pretty negligible, no particular need to optimize this yet.
+        # Maybe with a larger beam size...
+        finished_to_sort = [(-state.score.data[0] / len(state.action_history), state) for state in finished_states]
```

Comment:
I think this length normalisation is something that should happen in the `MaxMarginalLikelihood` decoder trainer as well?

Reply:
We talked about this in person, but to add my response here: `MaxMarginalLikelihood` doesn't have anything that would need this, but perhaps the `DecoderStep` should set the score to be normalized by number of steps.  It turns out that this isn't necessary if you're guaranteed to have every state be at the same timestep, which is currently the case, but I'll add a note and a TODO about this.